### ISSUE TITLE
`__based` is applied to every level of a multi-level pointer

### ISSUE THREAD

**Issue by WeiN76LQh** [Vector35/binaryninja-api#5896]
**Version and Platform (required):**
- Binary Ninja Version: 4.2.5996-dev (2f79eafb)
- OS: macOS
- OS Version: 14.5
- CPU Architecture: M2

**Bug Description:**
I think there's a bug with based pointers. If I create a type from C source, via the UI or the API and I specify a struct with a pointer like so:

```c
struct Foo {
    struct Bar ** __based(var) bar;
};
```

For some reason `Foo` always becomes:

```c
struct Foo
{
    struct Bar* __based(var)* __based(var) bar;
};
```

What I am declaring is a pointer which is a variable relative pointer and it points to a location in memory where there is an absolute pointer. However the type is getting converted to the destination type also being a variable relative pointer.

Doing the following with the API will result in the correct type being created so it seems its just an issue with the code that converts C source to types:

```python
bar = binaryninja.PointerBuilder.pointer_of_width(8, binaryninja.Type.pointer_of_width(8, bv.types["Bar"]))
bar.pointer_base_type = binaryninja.PointerBaseType.RelativeToVariableAddressPointerBaseType
bar.immutable_copy()
```

No matter how I write the pointer type I can't get it to end up being how I want it; a variable relative pointer to an absolute pointer to a struct of type `Bar`.

**Expected Behavior:**
`__based(var)` should not be applied to the 2nd level pointer unless specified like so:

```c
struct Foo {
    struct Bar * __based(var)* __based(var) bar;
};
```

**Additional Information:**
This does not appear to be a problem with `__ptr32`. So in the case of the following:

```c
struct Foo {
    struct Bar **__ptr32 bar;
};
```

The first level pointer will be 4 bytes in size and the second level pointer will be 8 bytes in size.

Doing the following:

```c
struct Foo {
    struct Bar * __ptr32* __ptr32 bar;
};
```

Results in both the first level and second level pointer being 4 bytes in size.

**Comment by CouleeApps** [comment]
Can confirm this is the case. This is probably due to the `__based` attribute being decl-based instead of type-based in our clang extensions. This should be pretty easy to fix now that our `handleTypeAttribute` patches have been implemented. Note to future self: Copy the implementation from `bn_ptr_width` and `bn_location` and mark the based status with an `AnnotateTypeAttr`.

**Comment by CouleeApps** [comment]
As a workaround for this for now, you can split each level of pointer in your type into a separate typedef and apply the annotations where they matter:

```
typedef Bar* pBar;
struct Foo {
    struct pBar * __based(var) bar;
};
```

**Comment by CouleeApps** [comment]
Fixed in 4.2.6242 as described above

**Comment by WeiN76LQh** [comment]
Hmm, it seems this change may have caused issues with `__ptr32` pointers but its hard to know if it was exactly this change as I can't go back to the version before this change on the development branch.

It seems a pointer with the attribute `__ptr32` will initially consume 8 bytes of space in a structure and then shrinks down to 4 bytes. This is the case when creating a new structure from C source but not the case when creating a structure with the API `StructureBuilder`.

Creating a type with the following declaration:
```c
struct Foo {
	void *__ptr32 bar;
};
```

Will result in a type in the types view that looks like this:
```c
struct Foo
{
    void* __ptr32 bar;
                  ?? ?? ?? ??
};
```

`bar` is the correct size, at 4 bytes but `Foo` is 8 bytes in size with an extra 4 bytes of unused space after `bar`. Adding the `__packed` attribute to `Foo` does fix it *in this case* but this didn't use to be what happened. It used to be that `Foo` would be 4 bytes without the `__packed` attribute.

Creating a type with the following declaration:
```c
struct Foo {
	uint32_t bar;
	void *__ptr32 baz;
};
```

Will result in a type in the types view that looks like this:
```c
struct Foo
{
    uint32_t bar;
    __offset(0x4);
    void* __ptr32 baz;
};
```

`Foo` is the correct size at 8 bytes and `bar` and `baz` are at their correct offsets (`0` and `4` respectively). However there is this odd `__offset(0x4);` line in there.

What I'm finding in practice is that larger structures containing `__ptr32` pointers are sometimes having the next field after the pointer be deleted and replaced with `?? ?? ?? ??`. The field after the deleted field is at the correct offset, which is why I say it seems these types are starting out as consuming 8 bytes in size, resulting in them trashing the next field, but then shrinking down to consume 4 bytes of space.

This is a problem on 4.2.6244-dev (c59961a8) and 4.2.6262-dev (5c271eff) but not on 4.2.6228-dev (312c3c99) and 4.1.5902 (f2165c5d).
